qiskit/algorithms/aux_ops_evaluator.py

```diff
@@ -0,0 +1,167 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021, 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""Evaluator of auxiliary operators for algorithms."""
+
+from typing import Tuple, Union, List
+
+import numpy as np
+
+from qiskit import QuantumCircuit
+from qiskit.algorithms.minimum_eigen_solvers.minimum_eigen_solver import ListOrDict
+from qiskit.opflow import (
+    CircuitSampler,
+    ListOp,
+    StateFn,
+    CircuitStateFn,
+    OperatorBase,
+    ExpectationBase,
+)
+from qiskit.providers import BaseBackend, Backend
+from qiskit.utils import QuantumInstance
+
+
+def eval_observables(
+    quantum_instance: Union[QuantumInstance, BaseBackend, Backend],
+    quantum_state: QuantumCircuit,
+    observables: ListOrDict[OperatorBase],
+    expectation: ExpectationBase,
+    threshold: float = 1e-12,
+) -> ListOrDict[Tuple[complex, complex]]:
+    """
+    Accepts a list or a dictionary of operators and calculates their expectation values - means
+    and standard deviations. They are calculated with respect to a quantum state provided. A user
+    can optionally provide a threshold value which filters mean values falling below the threshold.
+
+    Args:
+        quantum_instance: A quantum instance used for calculations.
+        quantum_state: An unparametrized quantum circuit representing a quantum state that
+            expectation values are computed against.
+        observables: A list or a dictionary of operators whose expectation values are to be
+            calculated.
+        expectation: An instance of ExpectationBase which defines a method for calculating
+            expectation values.
+        threshold: A threshold value that defines which mean values should be neglected (helpful for
+            ignoring numerical instabilities close to 0).
+
+    Returns:
+        A list or a dictionary of tuples (mean, standard deviation).
+
+    Raises:
+        ValueError: If a ``quantum_state`` with free parameters is provided.
+    """
+
+    if len(quantum_state.parameters) > 0:
+        raise ValueError(
+            "A parametrized quantum circuit representing a quantum_state was provided. It is not "
+            "allowed - the circuit cannot have free parameters."
+        )
+
+    # Create new CircuitSampler to avoid breaking existing one's caches.
+    sampler = CircuitSampler(quantum_instance)
+
+    list_op = _prepare_list_op(observables)
+
+    observables_expect = expectation.convert(
+        StateFn(list_op, is_measurement=True).compose(CircuitStateFn(quantum_state))
+    )
```

Comment:
Might need to handle more possible types of `quantum_state`. Do you have any suggestions?

Reply:
What other types are you considering and how do you anticipate computing expectation in these cases? For example in the numpy eigensolver https://github.com/Qiskit/qiskit-terra/blob/ec2a2a6c5b7e471e2f0479ba25af155516a63269/qiskit/algorithms/eigen_solvers/numpy_eigen_solver.py#L167 there is an equivalent for the state there, i.e a vector. Supporting other types etc whether it be state or operators seems to go somewhat beyond the original goal of simply being able to re-use what vqe did, which was with a circuit, and make it a public function that could be used elsewhere versus a private method. How does doing more read on Estimators - how much would we expect these to cover in that regard, since I thought this more of something that would just have a lifetime until algorithms were refactored to use those. Though from what I have seen so far these seem to be more tailored towards parameterized entities.

Reply:
Hmm, I did not notice that we also have a similar method in `numpy_eigen_solver.py`...
My comment originated from looking at what we have in the incoming Trotter algorithm, which will also use `aux_ops` as an option. What we have there is the following:
```
quantum_state = evolution_gate @ evolution_problem.initial_state
evolved_state = self._circuit_sampler.convert(quantum_state).eval()
```
which returns `VectorStateFn` as an `evolved_state` which then cannot be passed to the current code here because of the wrapping into `CircuitStateFn` on line 74 - it leads to an error. 
`TypeError: CircuitStateFn can only be instantiated with QuantumCircuit, not <class 'qiskit.opflow.state_fns.vector_state_fn.VectorStateFn'>`

So I temporarily added a fix like this:
```
    if isinstance(quantum_state, VectorStateFn):
        quantum_state = quantum_state.to_circuit_op()
    elif isinstance(quantum_state, QuantumCircuit):
        quantum_state = CircuitStateFn(quantum_state)

    observables_expect = expectation.convert(
        StateFn(list_op, is_measurement=True).compose(quantum_state)
    )
```.
So we could potentially cover some common types that might occur in this fashion...

Reply:
We might also do the foll.wing:
In `aux_ops_evaluator.py`:
```
observables_expect = expectation.convert(
        StateFn(list_op, is_measurement=True).compose(quantum_state)
    )
```
without adding `if` statements to check for instance type.
And in vqe.py:
```
aux_values = eval_observables(
                self.quantum_instance, CircuitStateFn(bound_ansatz), aux_operators, expectation=expectation
            )
```
instead of 
```
aux_values = eval_observables(
                self.quantum_instance, bound_ansatz, aux_operators, expectation=expectation
            )
```.
It works both for VQE unit tests and Trotter unit tests.

Reply:
Given Estimator takes circuits, and the intent would be to go over to using that, while thats what VQE uses today it would seem the Trotter algorithm could just convert its VectorStateFn to a circuit itself there so its always a circuit. This is what the Estimator interface looks like so it would be good to have whatever is done aligned to facilitate moving over to using these in the near future.

Reply:
I think it would be good if this would handle different types of state-functions so we can use it more generally. But we should avoid converting vectors to circuits, that's really expensive and inefficient. Instead, we could just have
```python
   observables_expect = expectation.convert(
         StateFn(list_op, is_measurement=True).compose(StateFn(quantum_state))
```
Opflow should take care of handling different types of `quantum_state` automatically. If `quantum_state` is a circuit it'll automatically construct a `CircuitStateFn`, if it's a vector a `VectorStateFn` etc.

Reply:
I guess I was thinking to avoid making it more general on the basis that I would expect VQE not to use this as soon as Estimators are done which would just leave the new Trotter algorithm using this. I am not sure whether it too can go over to using Estimators since from what I have seen the interface is circuit based hence the conversion suggestion. So I was more thinking this was an interim solution until that time. If its needed as a more general opflow oriented way of computing expectation using its primitives perhaps that valid, and perhaps it should be better located there.